test/Autofac.Test.Compilation/NullableReferenceTests.cs

```diff
@@ -0,0 +1,102 @@
+﻿using Xunit;
+#nullable enable
+
+namespace Autofac.Test.Compilation
+{
+    public class SimpleReferenceType
+    {
+    }
+
+    public class BaseClass
+    {
+
+    }
+
+    public class DerivedClass : BaseClass
+    {
+
+    }
+
+    public class NullableReferenceTests
+    {
+        [Fact]
+        public void NullableTypeNotAllowedForRegistration()
+        {
+            new AutofacCompile()
+                .Body(
+                @"
+                   var containerBuilder = new ContainerBuilder();
+                
+                   containerBuilder.Register<SimpleReferenceType?>(c => new SimpleReferenceType());
+                ")
+                .AssertWarningEndsWith("Nullability of type argument 'Autofac.Test.Compilation.SimpleReferenceType?' doesn't match 'notnull' constraint.");
+        }
+
+        [Fact]
+        public void NullableTypeNotAllowedForAsMethod()
+        {
+            var containerBuilder = new ContainerBuilder();
+
+            new AutofacCompile()
+                .Body(
+                @"
+                   var containerBuilder = new ContainerBuilder();
+                
+                   containerBuilder.Register(c => new DerivedClass()).As<BaseClass?>();
+                ")
+                .AssertWarningEndsWith("Nullability of type argument 'Autofac.Test.Compilation.BaseClass?' doesn't match 'notnull' constraint.");
```

Comment:
If we keep these tests and assert on the warning, we should only assert on things that aren't localizable like the names of types and constraints. Otherwise the tests will fail on a machine that has a different culture setting.

I'm OK if we don't keep these, or if we just keep it limited to one or two like this so things can be generally tested out. I can see a lot of these might become painful to maintain.

Reply:
I've amended the assertions to look for keywords only, that is a good tip (I did consider trying to force the compiler to invariant culture, but decided that really wasn't a great solution at all).